mbed_host_tests/host_tests/base_host_test.py

```diff
@@ -15,28 +15,116 @@
 limitations under the License.
 """
 
-class BaseHostTest():
+from time import time
+
+
+class BaseHostTestAbstract(object):
     """ Base class for each host-test test cases with standard
         setup, test and teardown set of functions
     """
 
     name = ''   # name of the host test (used for local registration)
+    event_queue = None      # To main even loop
+    dut_event_queue = None  # To DUT
 
-    def setup(self):
-        """ Set up function, initialize your test case dynamic resources in this function
+    def _notify_prn(self, text, nl=True):
+        if self.event_queue:
+            if nl:
+                text += '\n'
+            self.event_queue.put(('__notify_prn', text, time()))
+
+    def notify_complete(self, consume=True):
+        """! Notify htrun that host test finished processing
+        @param consume If True htrun will process (consume) all remaining events
         """
+        if self.event_queue:
+            self.event_queue.put(('__notify_complete', consume, time()))
+
+    def log(self, text):
+        self._notify_prn(text)
+
+    def _notify_dut(self, key, value):
+        """! Send data over serial to DUT """
+        if self.event_queue:
+            self.dut_event_queue.put((key, value, time()))
+
+    def send_kv(self, key, value):
+        self._notify_dut(key, value)
+
+    def setup(self):
+        pass
+
+    def test(self):
```

Comment:
What's the function of test() ? Is it required to be implemented or optional? If required, should it raise NotImplemented error here?